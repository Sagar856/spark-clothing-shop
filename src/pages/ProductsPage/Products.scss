.carousel-control-next-icon,
.carousel-control-prev-icon {
  height: 50px;
  width: 50px;
  outline: black;
  background-color: rgba(0, 0, 0, 0.3);
  background-size: 100%, 100%;
  border-radius: 50%;
}

.carousel-caption {
  background-color: rgb(86 77 77);
  padding-top: 0;
  padding-bottom: 0;
  border-radius: 15px;
}

.product-carousel-imgs {
  height: 400px;
  width: 1100px;
}

.rating {
  background: transparent;
  background-color: rgba(248, 250, 252, 0.795);
  font-size: 14px;
  font-weight: bold;
  padding: 5px;
  border-radius: 3px;
}

.category-btn-all {
  position: fixed;
  left: 805px;
  top: 87px;
  z-index: 1;
}

.category-btn-mens {
  position: fixed;
  left: 925px;
  top: 87px;
  z-index: 1;
  padding: 6px 16px;
}

.category-btn-womens {
  position: fixed;
  left: 1000px;
  top: 87px;
  z-index: 1;
  padding: 6px 17px;
}

.category-btn-kids {
  position: fixed;
  left: 1100px;
  top: 87px;
  z-index: 1;
}

.navigate-info {
  position: sticky;
  left: 695px;
  top: 87px;
  z-index: 1;
  border: none;
  background-color: white;
  color: red;
}
