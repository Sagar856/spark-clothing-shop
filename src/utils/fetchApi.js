export const fetchApi = (url, method) => {
  return fetch(url)
    .then((res) => {
      // if successful

      if (res.status == 200 || res.status == 201) {
        return res.json();
      } else {
        return {
          status: res.status,
          error: new Error("Invalid Response"),
        };
      }
    })
    .catch((err) => {
      // if error occurs
      return err;
    })
    .finally(() => {
    });
};
