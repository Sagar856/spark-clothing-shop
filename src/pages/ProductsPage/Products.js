import React, { useEffect, useState } from "react";
import { Helmet } from "react-helmet";
import { Link, Outlet } from "react-router-dom";
// Importing react icons
import { BsFillStarFill } from "react-icons/bs";
import { BsArrowRightCircleFill } from "react-icons/bs";
import { MdCelebration } from "react-icons/md";
import { fetchApi } from "../../utils/fetchApi";
import "./Products.scss";
import TabProducts from "../../components/TabProducts/TabProducts";

const Products = () => {
  const navigationBtns = [
    {
      id: 1,
      to: "/products",
      dataTestId: "allCategoryBtn",
      btnClassName: "btn btn-primary mx-4 category-btn-all",
      categoryName: "All",
      btnName: "All category",
    },
    {
      id: 2,
      to: "/products?category=mens",
      dataTestId: "mensCategoryBtn",
      btnClassName: "btn btn-primary mx-4 category-btn-mens",
      categoryName: "mens",
      btnName: "Men",
    },
    {
      id: 3,
      to: "/products?category=womens",
      dataTestId: "womensCategoryBtn",
      btnClassName: "btn btn-primary mx-4 category-btn-womens",
      categoryName: "womens",
      btnName: "Women",
    },
    {
      id: 4,
      to: "/products?category=kids",
      dataTestId: "kidsCategoryBtn",
      btnClassName: "btn btn-primary mx-4 category-btn-kids",
      categoryName: "kids",
      btnName: "Kids",
    },
  ];

  // Setting state using setter and getter Of useSate() hook
  const [products, setProducts] = useState();
  const [isError, setIsError] = useState(false);

  useEffect(() => {
    // Getting the products from API response
    fetchApi("http://localhost:5000/products")
      .then((res) => {
        setProducts(res);
      })
      // fetching error if any during loading data
      .catch((err) => {
        setIsError(true);
      });
  }, []);

  // Error message display
  if (isError) {
    return (
      <div className="alert alert-danger">
        Some Error Occurred! Try After Some Time
      </div>
    );
  }

  // Using handle for filtering products
  let url = null;
  const handleCategory = (category) => {
    // filtering products based on 'mens', 'womens', 'kids' category
    if (category !== "All") {
      // Using $ identifier to concatinate dynamic content
      url = `http://localhost:5000/products?category=${category}`;
    } else if (category === "All") {
      // filtering products for All categories
      url = "http://localhost:5000/products";
    }
    fetchApi(url).then((res) => {
      setProducts(res);
    });
  };

  return (
    <>
      {/* Using helmet to setting page title */}
      <Helmet>
        <title>Products</title>
      </Helmet>
      <div className="container">
        <div className="row">
          <div className="text-center"></div>
          <div className="container mx-2">
            <div className="row mt-4 mx-2">
              <div className="col-md-9">
                {/* Category wise filtering buttons */}
                {navigationBtns.map(
                  ({
                    id,
                    to,
                    dataTestId,
                    btnClassName,
                    categoryName,
                    btnName,
                  }) => (
                    <Link to={to} key={id}>
                      <button
                        data-testid={dataTestId}
                        className={btnClassName}
                        onClick={() => handleCategory(categoryName)}
                      >
                        {btnName}
                      </button>
                    </Link>
                  )
                )}
                {/* using react icon */}
                <button className="navigate-info">
                  Navigate here <BsArrowRightCircleFill />
                </button>
              </div>
            </div>

            <div className="row">
              {/* Rendering TabProducts, i.e carousel */}
              <TabProducts />
              <h3 className="my-3">
                {/* React icon */}
                DEALS OF THE DAY <MdCelebration />
              </h3>
              {/* Mapping product details using map() function */}
              {products &&
                products.map(
                  ({
                    id,
                    img,
                    title,
                    description,
                    offerPrice,
                    actualPrice,
                    discountPercent,
                    rating,
                    reviews,
                  }) => {
                    return (
                      <div className="col-md-3" key={id}>
                        {/* Navigating to product detailes by id on click of the product */}
                        <Link
                          to={`/products/${id}`}
                          className="text-dark text-decoration-none"
                        >
                          <div className="card my-2 shadow-sm border-removal">
                            <div className="position-relative ">
                              <img
                                className="w-100 h-75"
                                src={img}
                                alt={title}
                              />
                              <div className="position-absolute bottom-0">
                                <p className="p-1 m-1 rating text-dark">
                                  {rating}
                                  <BsFillStarFill />
                                  {reviews}
                                </p>
                              </div>
                            </div>
                            <div className="card-body">
                              <h5 className="card-title text-start fw-bold">
                                {title}
                              </h5>
                              <p className="card-text text-secondary text-start">
                                {description}
                              </p>
                              <div className="text-start">
                                <span className="me-2 fw-bold">
                                  Rs.{offerPrice}
                                </span>
                                <span className="me-3 text-secondary text-decoration-line-through">
                                  Rs.{actualPrice}
                                </span>
                                <span className="me-2 text-danger">
                                  ({discountPercent}% OFF)
                                </span>
                              </div>
                            </div>
                          </div>
                        </Link>
                      </div>
                    );
                  }
                )}
            </div>
          </div>
        </div>
      </div>
      {/* Outlet for arranging the stuff */}
      <Outlet />
    </>
  );
};

export default Products;
