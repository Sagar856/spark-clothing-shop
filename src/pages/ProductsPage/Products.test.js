import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { HashRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import { fetchApi } from "../../utils/fetchApi";
import Products from "./Products";

// setting up mock for fetchApi
// mocking a module with an automocked version when needed
jest.mock("../../utils/fetchApi");

describe("products", () => {
  it("renders error properly during API call", async () => {
    fetchApi.mockRejectedValue({
      errorMessage: "Some Error Occurred! Try After Some Time",
    });
    render(
      <HashRouter>
        <Products />
      </HashRouter>
    );
    expect(
      await screen.findByText("Some Error Occurred! Try After Some Time")
    ).toBeInTheDocument();
  });

  it("fetches Products properly via API call [MOCKING]", async () => {
    // prepare the mock data for the users
    const mockProduct = [
      {
        id: 1,
        description: "Men Pure Cotton Casual Shirt",
        title: "Roadster",
        img: "/assets/images/Product1.png",
        offerPrice: "674",
        actualPrice: "1499",
        discountPercent: "55",
        material: "100% Cotton",
        care: "Machine-wash",
      },
    ];
    fetchApi.mockResolvedValue(mockProduct);
    // render the comp
    render(
      <HashRouter>
        <Products />
      </HashRouter>
    );
    //  then, assert
    const descriptionEl = await screen.findByText(
      /Men Pure Cotton Casual Shirt/i
    );
    expect(descriptionEl).toBeInTheDocument();
    await expect(screen.getByText(/Roadster/i)).toBeInTheDocument();
    await expect(
      screen.getByText(/Men Pure Cotton Casual Shirt/i)
    ).toBeInTheDocument();
    await expect(screen.getByText(/674/i)).toBeInTheDocument();
    await expect(screen.getByText(/1499/i)).toBeInTheDocument();
    await expect(screen.getByText(/55/i)).toBeInTheDocument();
  });

  it("has proper working category buttons as 'All category'", async () => {
    fetchApi.mockResolvedValue();

    render(
      <HashRouter>
        <Products />
      </HashRouter>
    );
    const allCategoryBtn = await screen.getByTestId("allCategoryBtn");
    expect(allCategoryBtn).toBeInTheDocument();
    expect(allCategoryBtn).toHaveTextContent("All category");
  });

  // Snapshot testing
  it("has right snapshot with all requirements completed", () => {
    // to take snapshot we need react-test-renderer.
    const snapshotInJson = renderer
      .create(
        <HashRouter>
          <Products />
        </HashRouter>
      )
      .toJSON();
    // let assert with a matchrer toMatchSnapshot()
    expect(snapshotInJson).toMatchSnapshot();
  });

  it("should render the available products properly", async () => {
    fetchApi.mockResolvedValue();
    render(
      <HashRouter>
        <Products />
      </HashRouter>
    );

    const mensCategoryBtn = await screen.findByTestId("mensCategoryBtn");
    const womensFashion = await screen.findByTestId("womensCategoryBtn");
    const kidsFashion = await screen.findByTestId("kidsCategoryBtn");
    const allCategoryBtn = await screen.findByTestId("allCategoryBtn");

    expect(allCategoryBtn.textContent).toBe("All category");

    await waitFor(() => {
      fireEvent.click(mensCategoryBtn);
      const dnone = screen.getByTestId("mensCategoryBtn");
      expect(dnone.textContent).toBe("Men");
    });

    await waitFor(() => {
      fireEvent.click(womensFashion);
      const womensCategoryBtn = screen.getByTestId("womensCategoryBtn");
      expect(womensCategoryBtn.textContent).toBe("Women");
    });

    await waitFor(() => {
      fireEvent.click(kidsFashion);
      const kidsCategoryBtn = screen.getByTestId("kidsCategoryBtn");
      expect(kidsCategoryBtn.textContent).toBe("Kids");
    });
  });
});
