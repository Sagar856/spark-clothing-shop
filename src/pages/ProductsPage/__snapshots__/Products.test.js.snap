// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`products has right snapshot with all requirements completed 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="text-center"
    />
    <div
      className="container mx-2"
    >
      <div
        className="row mt-4 mx-2"
      >
        <div
          className="col-md-9"
        >
          <a
            href="#/products"
            onClick={[Function]}
          >
            <button
              className="btn btn-primary mx-4 category-btn-all"
              data-testid="allCategoryBtn"
              onClick={[Function]}
            >
              All category
            </button>
          </a>
          <a
            href="#/products?category=mens"
            onClick={[Function]}
          >
            <button
              className="btn btn-primary mx-4 category-btn-mens"
              data-testid="mensCategoryBtn"
              onClick={[Function]}
            >
              Men
            </button>
          </a>
          <a
            href="#/products?category=womens"
            onClick={[Function]}
          >
            <button
              className="btn btn-primary mx-4 category-btn-womens"
              data-testid="womensCategoryBtn"
              onClick={[Function]}
            >
              Women
            </button>
          </a>
          <a
            href="#/products?category=kids"
            onClick={[Function]}
          >
            <button
              className="btn btn-primary mx-4 category-btn-kids"
              data-testid="kidsCategoryBtn"
              onClick={[Function]}
            >
              Kids
            </button>
          </a>
          <button
            className="navigate-info"
          >
            Navigate here 
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 0a8 8 0 1 1 0 16A8 8 0 0 1 8 0zM4.5 7.5a.5.5 0 0 0 0 1h5.793l-2.147 2.146a.5.5 0 0 0 .708.708l3-3a.5.5 0 0 0 0-.708l-3-3a.5.5 0 1 0-.708.708L10.293 7.5H4.5z"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        className="row"
      >
        <div>
          <hr />
          <div
            className="carousel slide carousel-resizing"
            data-testid="tabCarouselElement"
            id="carouselExampleCaptions"
          >
            <div
              className="carousel-inner"
            >
              <div
                className="carousel-item"
              >
                <img
                  alt="KidsWearImg"
                  className="d-block product-carousel-imgs"
                  src="/assets/images/carousel/KidsWearImg.png"
                />
              </div>
              <div
                className="carousel-item"
              >
                <img
                  alt="WomenWearImg"
                  className="d-block product-carousel-imgs"
                  src="/assets/images/carousel/WomenWearImg.png"
                />
              </div>
              <div
                className="carousel-item active"
              >
                <img
                  alt="MensWearImg"
                  className="d-block product-carousel-imgs"
                  src="/assets/images/carousel/MensWearImg.png"
                />
              </div>
            </div>
            <div>
              <button
                className="carousel-control-prev"
                data-bs-slide="prev"
                data-bs-target="#carouselExampleCaptions"
                type="button"
              >
                <span
                  aria-hidden="true"
                  className="carousel-control-prev-icon"
                />
                <span
                  className="visually-hidden"
                >
                  Previous
                </span>
              </button>
              <button
                className="carousel-control-next"
                data-bs-slide="next"
                data-bs-target="#carouselExampleCaptions"
                type="button"
              >
                <span
                  aria-hidden="true"
                  className="carousel-control-next-icon"
                />
                <span
                  className="visually-hidden"
                >
                  Next
                </span>
              </button>
            </div>
          </div>
        </div>
        <h3
          className="my-3"
        >
          DEALS OF THE DAY 
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M2 22l14-5-9-9zM14.53 12.53l5.59-5.59a1.25 1.25 0 011.77 0l.59.59 1.06-1.06-.59-.59a2.758 2.758 0 00-3.89 0l-5.59 5.59 1.06 1.06zM10.06 6.88l-.59.59 1.06 1.06.59-.59a2.758 2.758 0 000-3.89l-.59-.59-1.06 1.07.59.59c.48.48.48 1.28 0 1.76zM17.06 11.88l-1.59 1.59 1.06 1.06 1.59-1.59a1.25 1.25 0 011.77 0l1.61 1.61 1.06-1.06-1.61-1.61a2.758 2.758 0 00-3.89 0zM15.06 5.88l-3.59 3.59 1.06 1.06 3.59-3.59a2.758 2.758 0 000-3.89l-1.59-1.59-1.06 1.06 1.59 1.59c.48.49.48 1.29 0 1.77z"
            />
          </svg>
        </h3>
      </div>
    </div>
  </div>
</div>
`;
